### PR TITLE
Fix electron issues

### PR DESCRIPTION
**Pull request by edvald**
This aims to address the issues in #59 and #123. I'll note that personally I am not using Electron, but came across these issues when making a bundle via zeit/pkg. So it would be great if someone can test these changes in Electron directly. I also don't directly use any code that touches on the WebSocketHandler, so someone might want to do additional testing there.

Further, I looked into getting rid of synchronous execs and file reads but decided not to include those changes in here, because they would introduce API changes in various places (basically making a bunch of surface-level methods async that are currently synchronous). @brendandburns I'd highly recommend making those changes since I kinda think they'll be necessary at some point anyway, and I'm happy to make a PR for that, but I wanted to get your input first.

### CODE REVIEW

**Comment by edvald** [comment]
Seems the Github downtime is causing issues here - I've updated my branch to fix the CI errors, but it's not showing up here yet. Figure it'll snap into place when Github is back in shape.

**Comment by itowlson** [comment]
+1 to providing async options for things involving execs and reads, but also +1 to not breaking the existing API surface!  Should we open an issue so that we can discuss the way forward on this?

File: node-client/src/config.ts [hunk 1]
```diff
@@ -273,12 +273,10 @@ export class KubeConfig {
...
                     if (config['cmd-path']) {
-                        let cmd = '"' + config['cmd-path'] + '"';
-                        if (config['cmd-args']) {
-                            cmd = cmd + ' ' + config['cmd-args'];
-                        }
+                        const args = config['cmd-args'] ? [config['cmd-args']] : [];
                         // TODO: Cache to file?
-                        const result = shelljs.exec(cmd, { silent: true });
+                        // TODO: do this asynchronously
+                        const result = execa.sync(config['cmd-path'], args);
```
Comment (itowlson) [inline node-client/src/config.ts]: I don't know the inner workings of `shelljs.exec` but the name implies it used to go via the shell.  Should we therefore use `execa.shellSync`?  (This is a question, not a suggestion... I lack knowledge and would value your advice.)

File: node-client/src/web-socket-handler.ts [hunk 2]
```diff
@@ -96,37 +104,38 @@ export class WebSocketHandler implements WebSocketInterface {
...
+                }
+            };
+
+            client.onmessage = ({ data }: { data: WebSocket.Data }) => {
+                // TODO: support ArrayBuffer and Buffer[] data types?
+                if (typeof data === 'string') {
+                    if (textHandler && !textHandler(data)) {
+                        client.close();
+                    }
+                } else if (binaryHandler && data instanceof Buffer) {
```
Comment (itowlson) [inline node-client/src/web-socket-handler.ts]: I'd suggest we keep the existing structure where the top level `if` checks only on the type of data, and the second level makes it a no-op if there is no handler for the type of data received:

```
if (data instanceof Buffer) {
  if (binaryHandler) {
    // ...
  }
}
```

This keeps the structure consistent with the string case (although of course at the moment the behaviour is the same either way).

**Comment by edvald** [comment]
@mkramb could you test this to see if it addresses your issue?

**Comment by edvald** [comment]
Oh and also @cabrinoob, would be great to know if this fixes #123 for you

**Comment by cabrinoob** [comment]
Hi, 
I've just pull the PR changes and integrated them in my electron App to check if it fixes the #123 issue. I still have an error but with a different message.

``` 
Error: Uncaught (in promise): Error: ws does not work in the browser. Browser clients must use the native WebSocket object
Error: ws does not work in the browser. Browser clients must use the native WebSocket object
    at new push../node_modules/@kubernetes/client-node/node_modules/ws/browser.js.module.exports (browser.js:4)
    at Promise (web-socket-handler.js:79)
```

**Comment by edvald** [comment]
Ah, so we'll need to go one step further and set up https://github.com/heineiuo/isomorphic-ws. I'm flying a little blind here, but I've added a commit for that. Let me know if it helps.

File: node-client/src/config_test.ts [hunk 3]
```diff
@@ -535,8 +535,7 @@ describe('KubeConfig', () => {
                     },
                 } as User);
             const opts = {} as requestlib.Options;
-            expect(() => config.applyToRequest(opts)).to.throw(
-                'Failed to refresh token: /bin/sh: 1: non-existent-command: not found');
+            expect(() => config.applyToRequest(opts)).to.throw('non-existent-command');
```
Comment (brendandburns) [inline node-client/src/config_test.ts]: Let's keep the 'failed to refresh token' part...

**Comment by brendandburns** [comment]
@edvald I merged your other PR so this one now has conflicts...

**Comment by drubin** [comment]
@brendanburns Is it possible to release a patch for this? Or will it have to wait for 0.8?

**Pull request merged**